amethyst_tiles/src/pass.rs

```diff
@@ -0,0 +1,451 @@
+    Z: DrawTiles2DBounds = DrawTiles2DBoundsDefault,
+> {
+    pipeline: B::GraphicsPipeline,
+    pipeline_layout: B::PipelineLayout,
+    textures: TextureSub<B>,
+    vertex: DynamicVertexBuffer<B, TileArgs>,
+    sprites: OrderedOneLevelBatch<TextureId, TileArgs>,
+    change: util::ChangeDetection,
+
+    env: DynamicUniform<B, TileMapArgs>,
+
+    #[derivative(Debug = "ignore")]
+    _marker: PhantomData<(T, E, Z)>,
+}
+
+impl<B: Backend, T: Tile, E: CoordinateEncoder, Z: DrawTiles2DBounds> RenderGroup<B, World>
+    for DrawTiles2D<B, T, E, Z>
+{
+    #[allow(clippy::cast_precision_loss)]
+    fn prepare(
+        &mut self,
+        factory: &Factory<B>,
+        _queue: QueueId,
+        index: usize,
+        _subpass: hal::pass::Subpass<'_, B>,
+        world: &World,
+    ) -> PrepareResult {
+        #[cfg(feature = "profiler")]
+        profile_scope!("prepare");
+
+        let mut changed = false;
+        let (sprite_sheet_storage, tex_storage, hiddens, tile_maps, transforms) =
+            <(
+                Read<'_, AssetStorage<SpriteSheet>>,
+                Read<'_, AssetStorage<Texture>>,
+                ReadStorage<'_, Hidden>,
+                ReadStorage<'_, TileMap<T, E>>,
+                ReadStorage<'_, Transform>,
+            )>::fetch(world);
+
+        let sprites_ref = &mut self.sprites;
+        let textures_ref = &mut self.textures;
+
+        sprites_ref.swap_clear();
+
+        let CameraGatherer { projview, .. } = CameraGatherer::gather(world);
+
+        let mut tilemap_args = TileMapArgs {
+            proj: projview.proj,
+            view: projview.view,
+            map_coordinate_transform: Default::default(),
+            map_transform: Default::default(),
+            sprite_dimensions: Default::default(),
+        };
+
+        for (tile_map, _, transform) in (&tile_maps, !&hiddens, transforms.maybe()).join() {
+            let map_coordinate_transform: [[f32; 4]; 4] = (*tile_map.transform()).into();
+
+            let map_transform: [[f32; 4]; 4] = if let Some(transform) = transform {
+                (*transform.global_matrix()).into()
+            } else {
+                Matrix4::identity().into()
+            };
+
+            tilemap_args.map_coordinate_transform = map_coordinate_transform.into();
+            tilemap_args.map_transform = map_transform.into();
+            tilemap_args.sprite_dimensions = [
+                tile_map.tile_dimensions().x as f32,
+                tile_map.tile_dimensions().y as f32,
+            ]
+            .into();
+
+            let region = Z::bounds(tile_map, world);
+            region
+                .iter()
+                .filter_map(|coord| {
+                    let tile = tile_map.get(&coord).unwrap();
+                    if let Some(sprite_number) = tile.sprite(coord, world) {
+                        let (batch_data, texture) = {
+                            if let Some(sprite_sheet_handle) = tile_map.sprite_sheet.as_ref() {
```

Comment:
this whole `sprite_sheet_handle` dance here can be extracted to the outer loop, just below `for (tile_map, _, transform)`.

```rust
let maybe_sheet = tile_map
	.sprite_sheet.map(|handle| sprite_sheet_storage.get(handle))
	.filter(|sheet| tex_storage.contains(&sheet.texture));
	
let sprite_sheet = match maybe_sheet {
  Some(sheet) => sheet,
  None => continue,
}
```